src/lualib/Decorate.ts

```diff
@@ -0,0 +1,17 @@
+/**
+ * NOTE: will need to be changed to support Method and Property decoration
+ *
+ * SEE: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/transformers/ts.ts#L3598
+ */
+function __TS__Decorate(this: void, decorators: Function[], target: {}, key?: string, desc?: any): {} {
+    let r = target;
+
+    for (let i = decorators.length; i >= 0; i--) {
+        const d = decorators[i];
+        if (d) {
+            r = (key === undefined ? d(r) : desc !== undefined ? d(target, key, r) : d(target, key)) || r;
```

Comment:
I think there's no much reason to have helpers be 1:1 with TS ones - that condition generates quite unoptimized code with few IIFEs and variables may be renamed for readability, like:

```ts
function __TS__Decorate(this: void, decorators: Function[], target: {}, key?: string, desc?: any): {} {
    let result = target;

    for (let i = decorators.length; i >= 0; i--) {
        const decorator = decorators[i];
        if (decorator) {
            const oldResult = result;

            if (key === undefined) {
                result = decorator(result);
            } else if (desc !== undefined) {
                result = decorator(target, key, result);
            } else {
                result = decorator(target, key);
            }

            result = oldResult || result;
        }
    }

    return result;
}
```

Reply:
True, this may look like optimized js, but it does not necessarily lead to optimal lua.

Reply:
Note that in ark's code I think he meant the `result =` assignments to be `oldResult =` in the if statement.

Reply:
Oh, I meant `result = result || oldResult`, but either one would work (though naming would be a little weird if `oldResult` was reassigned)